week7/cooper_client_react/04_saving_data.md

```diff
@@ -0,0 +1,252 @@
+
+```js
+  <DisplayCooperResult
+    distance={this.state.distance}
+    gender={this.state.gender}
+    age={this.state.age}
+    authenticated={this.state.authenticated}
+    entrySaved={this.state.entrySaved}
+    entryHandler={this.entryHandler.bind(this)}
+  />
+```
+
+If you run the feature test now, you will get an error that says "Text not found "Your entry was saved". That's not weird, we haven't rendered the message we get from the backend anywhere. If you take a look at logs for the backend, you will see that the performance data is being saved correctly. We need to add and modify some of our code to get it to go green.
+
+First we need to modify the if statement in `DisplayCooperResult` component:
+
+```js
+render() {
+  if (this.props.authenticated === true && this.props.entrySaved === false) {
+    saveButton = (
+      <>
+        <button id="save-result" onClick={this.saveCooperData.bind(this)}>Save entry</button>
+      </>
+    )
+  } else if (this.props.authenticated === true && this.props.entrySaved === true) {
+    saveButton = (
+      <>
+        <p>Your entry was saved</p>
+      </>
+    )
+  }
+  if (this.props.age !== '' && this.props.distance !== '') {
+    results = (
+    // ...
+    )
+  }
+  // ...
+}
+```
+
+Previously we modified the rendering of this component in the `App` component to pass in the state of `entrySaved`. So if we have saved the result we want to see `"Your entry was saved"`. Now we need to modify the `onChange` function in the app component so that every time we modify the inputs we can save a new result. 
+
+```js
+  onChange(event) {
+    this.setState({
+      [event.target.id]: event.target.value,
+      entrySaved: false
+    })
+  }
+```
+
+If you run the test now, both test in the feature file should go green.
+
+We want to be able to run these test without sending requests to the backend. Let's mock the response out. 
+
+First we need to add some code to the `mocksConfig` file, so it intercepts the `/performance_data` request:
+
+```js  
+const createResponse = (path, params, request) => {
+    let response
+    switch (path) {
+      case 'sign_in':
+        let user
+        user = MockResponses.mockedUserResponses.find(user => {
+          return user.headers.uid === JSON.parse(params).email
+        })
+        response = user || MockResponses.missingUserResponse
+        return response
+      case 'performance_data':
+        return MockResponses.savingEntryResponse
+    }
+    
+  }
+
+  const requests = {
+    'sign_in': {},
+    'performance_data': {}
+  }
+
+  // ...
+```
+
+Now we need to add the response that we are calling on when puppeteer intercepts `/performance_data`
+
+Add this to `mockResponses`:
+
+```js
+// ...
+  savingEntryResponse: {
+    status: 200,
+    headers: {},
+    body: JSON.stringify({
+      message: "all good"
+    })
+  }
+```
+
+You also need to require the `mocksConfig` file to the `userCanSavePerformanceData.feature.js` feature test.
```

Comment:
We, instead of you to keep it consistent

Reply:
done